### PR TITLE
Follow unfollow asynchronous requests

### LINKED ISSUE #3559: Improve Follow/unfollow button
### Please describe the problem (or idea)
In user profile page if we want to follow to that user then click on follow button but after click page getting refreshing so need to add ajax here
> What happened just before the problem occurred? Or what problem could this idea solve?

> What did you expect to see that you didn't?

page refreshing after click on follow/unfollow button

### Please show us where to look

https://publiclab.org/profile

### What's your PublicLab.org username?
singhav
> This can help us diagnose the issue: 

### Browser, version, and operating system

> Many bugs are related to these -- please help us track it down and reproduce what you're seeing!

****

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site. 

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders

### PR DESCRIPTION
**Pull request by shubhscoder**
Fixes #3559 
This PR proposes the following changes-
1. You can now follow a user without refreshing the page, also the follow button changes to unfollow upon successful following with a alert regarding the same.
2. You can now unfollow a user without refreshing the page, also the unfollow button changes to follow upon successful unfollowing with a alert regarding the same.

Thank you!

### CODE REVIEW

**Comment by shubhscoder** [comment]
@gauravano the follow request made from profile.html is always going as a get request despite doing
<code> method = "post" </code> . I tried googling and did everything I could, but could not figure that out. Do you have any idea why this is happening?

**Comment by gauravano** [comment]
Will see this in some hours @shubhscoder. Also, there are some merge conflicts, how about resolving those first? Thanks!

**Comment by shubhscoder** [comment]
@gauravano @jywarren I think this is ready for review. Thank you so much!

File: test/functional/relationships_controller_test.rb [hunk 1]
```diff
@@ -23,11 +22,9 @@ def setup
     post :create, params: { followed_id: followed_user.id }
 
     assert_difference 'Relationship.count', -1 do
-      delete :destroy, params: { id: Relationship.last.id }
+      delete :destroy, params: { id: followed_user.id }
     end
-
-    assert_response :redirect
-    assert_redirected_to '/profile/' + followed_user.username
```
Comment (jywarren) [inline test/functional/relationships_controller_test.rb]: Hi! Should we test the Json response itself?

File: app/views/users/profile.html.erb [hunk 2]
```diff
@@ -341,6 +341,25 @@
...
+    $('#follow_form').html('<a id="profile-unfollow-button" class="btn btn-default" data-method="delete" data-remote="true" href="/relationships/<%=@user.id%>">Unfollow</a>');
+  });
+  $('#follow_form').on('ajax:success', '#profile-unfollow-button', function(data, status, xhr)
+  {
+    notyNotification('relax', 3000, 'success', 'top', 'You unfollowed <%=@user.username%>');
+    $('#follow_form').html('<a id="profile-follow-button" class="btn btn-primary" data-method="post" data-remote="true" href="/relationships" data-params="followed_id=<%=@user.id%>">Follow</a>');
+  });
+  $('#follow_form').on('ajax:error', '#profile-follow-button', function(data, status, xhr)
+  {
+    notyNotification('relax', 3000, 'error', 'top', 'Error in following');
```
Comment (gauravano) [inline app/views/users/profile.html.erb]: How about Error in following the user or while following the user?

File: app/views/users/profile.html.erb [hunk 3]
```diff
@@ -341,6 +341,25 @@
...
+  });
+  $('#follow_form').on('ajax:error', '#profile-unfollow-button', function(data, status, xhr)
+  {
+    notyNotification('relax', 3000, 'error', 'top', 'Error in following');
```
Comment (gauravano) [inline app/views/users/profile.html.erb]: this should be unfollowing? right? Also, change this message to follow the pattern if you make this change

**Comment by gauravano** [comment]
And sorry @shubhscoder, I committed https://github.com/publiclab/plots2/pull/4439#issuecomment-470998332  but wasn't able to look earlier.

**Comment by gauravano** [comment]
Hey @shubhscoder, I think https://github.com/publiclab/plots2/blob/master/app/views/like/_like.html.erb#L12-L15 requires some changes as the follow button is behaving like 

follow_unfollow

Also, as @jywarren said, we can test for both JSON response as well as DB entry. 

Thanks!

**Comment by shubhscoder** [comment]
> And sorry @shubhscoder, I committed [#4439 (comment)](https://github.com/publiclab/plots2/pull/4439#issuecomment-470998332) but wasn't able to look earlier.

No problem @gauravano !

**Comment by shubhscoder** [comment]
> Hey @shubhscoder, I think https://github.com/publiclab/plots2/blob/master/app/views/like/_like.html.erb#L12-L15 requires some changes as the follow button is behaving like
> 
> follow_unfollow
> 
> Also, as @jywarren said, we can test for both JSON response as well as DB entry.
> 
> Thanks!

@gauravano this is related to #4587 right? I will fix that as soon as possible. Thank you for letting me know that,

**Comment by shubhscoder** [comment]
Do we need any more changes in this @jywarren @gauravano ?

**Comment by gauravano** [comment]
@shubhscoder no this- https://github.com/publiclab/plots2/issues/4991 is related to follow user only. We have option to follow user there too, see the code link. Thanks!

**Comment by shubhscoder** [comment]
@gauravano I realised what is happening. The data-method = "post" is somehow not working in _like partial. That is a issue on production as well. We are redirected to tags/relationships if we try to follow someone through the _like partial. This is because somehow it is being interpreted as a get request and not a post request and thus the resources are not working. Do you what could be reason for the data-method = 'post' not working? 
(You can check the error on production here : https://publiclab.org/notes/rexagod/03-11-2019/gsoc-proposal-mapknitter-orb-descriptor-w-auto-stitching-pattern-training-and-live-video-support-and-ldi-revamp-major-ui-enhancements) You can try following rexagod, you will be redirected to tags/realtionships

**Comment by SidharthBansal** [comment]
There are some conflicts. Can you please resolve them?
Is there any progress on this issue guys?

**Comment by gauravano** [comment]
Ok, I have removed the conflict and now seeing code

**Comment by gauravano** [comment]
Hi @jywarren @publiclab/plots2-reviewers, I have did some re-writing and I think it's ready. A user can now be followed from the profile as well as a note page.
 
follow_work1

Also, exception handling is working fine too 


ezgif com-optimize (4)

Let me know if anything else is required. Thanks!!

**Comment by gauravano** [comment]
Hey @jywarren, please review this PR.

**Comment by jywarren** [comment]
This looks great; merging. It may be a good candidate for a system test as well!

**Pull request merged**